File: cluster/node-up.rb
require 'shellwords'

# Assuming we're working with a fresh vm...

class ComputeInstanceBuilder
  attr_reader :username

  def initialize(cloud_name, image_id, flavor_id, ssh_key, security_groups, username)
    @cloud_name = Shellwords.escape(cloud_name)
    @image_id = Shellwords.escape(image_id)
    @flavor_id = Shellwords.escape(flavor_id)
    @ssh_key = Shellwords.escape(ssh_key)
    @security_groups = security_groups.map { |key| Shellwords.escape(key) }
    @username = username
  end

  def build(name)

    security_groups = @security_groups
      .map { |group| "--security-group #{group}" }
      .join(' ')
    
    puts "creating #{name}"
    `openstack --os-cloud #{@cloud_name} server create \
      --wait \
      --image #{@image_id} \
      --flavor #{@flavor_id} \
      --key-name #{@ssh_key} \
      #{security_groups} \
      #{name}
    `

    return ComputeInstance.new(self, name)
  end

end
# ansible_user=debian ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_python_interpreter=python3
class ComputeInstance
  attr_reader :name
  attr_reader :builder
  attr_accessor :ip

  def initialize(builder, name)
    @builder = builder
    @name = name
  end

  def append_to_hostfile(writer, keyfile)
    writer.puts("#{@ip} ansible_user=#{@builder.username} ansible_ssh_private_key_file=#{keyfile} ansible_python_interpreter=python3")
  end

  def self.write_hostfile(instances, keyname, path)
    puts "Path to private key for keypair #{keyname}: "
    keypath = gets.strip
    open(path, 'w') { |f|
      f.puts('[manager]')
      instances[0].append_to_hostfile(f, keypath)
      f.puts('[worker]')
      instances[1..].each { |i| i.append_to_hostfile(f, keypath) }
    }
  end

  def self.populate_ips(instances, cloud_name)
    # Get server info from openstack
    host_names = instances.map { |i| "--host #{Shellwords.escape(i.name)}" }.join(' ')
    # Columns are <empty>, id, name, status, networks, image, flavor
    ips = `openstack --os-cloud #{cloud_name} server list #{host_names}`
      .lines
      .filter_map { |l|
        cols = l.split('|')
        [cols[2].strip, cols[4].strip] if cols.length > 4
      }
    puts ips.inspect
    ips = ips
      .filter_map { |l|
        networks_split = l[1].strip.split('=')
        if networks_split.length < 2 then
          nil
        else
          l[1] = networks_split[1]
          l
        end
      }
      .to_h
    puts ips.inspect
    instances.each { |i|
      puts "Found ip: #{i.name} -> #{ips[i.name]}"
      i.ip = ips[i.name]
    }
  end

  def self.ansible_up(instances, hostfile)
    system("ansible-playbook -i #{hostfile} ansible/bootstrap.yml")
  end
end

builder = ComputeInstanceBuilder.new(
  'nectar-private',
  'NeCTAR Debian 10 (Buster) amd64',
  't3.xsmall',
  'alaroldai_haliax',
  ['default', 'ssh', 'mosh', 'icmp'],
  'debian'
)

# instances = ['test', 'test2']
#   .map { |name| builder.build(name) } # -> [ComputeInstance]
instances = [
  ComputeInstance.new(builder, 'test'),
  ComputeInstance.new(builder, 'test2'),
]
ComputeInstance.populate_ips(instances, 'nectar-private')
ComputeInstance.write_hostfile(instances, 'alaroldai_haliax', 'ansible/generated_hosts.ini')
ComputeInstance.ansible_up(instances, 'ansible/generated_hosts.ini')

File: cluster/node-up.py
import asyncio
import json
from getpass import getpass
from sys import argv
from os import environ
from pprint import pprint
import subprocess


from log import *

class OpenStackClient:
  def __init__(self, cloud_name):
    self.__cloud_name = cloud_name
    self.__password = None
    self.__known_stacks = None

  @property
  def password(self):
    if not self.__password:
      self.__password = environ['OS_PASSWORD'] if 'OS_PASSWORD' in environ else None
    if not self.__password:
      self.__password = getpass(f"Enter password for {self.__cloud_name}: ")
    return self.__password


  async def stack_exists(self, name):
    if not self.__known_stacks:
      trace('Fetching stack list')
      cmd = [
      'openstack',
      '--os-cloud', self.__cloud_name,
      'stack', 'list',
      '-f', 'json'
      ]
      proc = await asyncio.create_subprocess_exec(
        *cmd,
        env=environ.copy() | { 'OS_PASSWORD' : self.password },
        stdout=asyncio.subprocess.PIPE
      )
      stdout, _ = await proc.communicate()
      stdout = stdout.decode()

      js = json.loads(stdout)
      self.__known_stacks = { s['Stack Name'] : s for s in js }
    pprint(self.__known_stacks)
    return name in self.__known_stacks

  async def create_stack(self, template, name, params):
    action = 'create'
    if await self.stack_exists(name):
      action = 'update'
      if input('Stack already exists, update it? (y/n): ') != 'y':
        return
    trace(f'{"updating" if action == "update" else "creating"} stack {name}')
    cmd = [
      'openstack',
      '--os-cloud', self.__cloud_name,
      'stack', action,
      '-f', 'json',
      '--wait',
      '-t', template,
      name
    ] + [i for key, value in params.items() for i in ['--parameter', f'{key}={value}']]
    trace('Calling for effect: ' + ' '.join(cmd))
    proc = await asyncio.create_subprocess_exec(
      *cmd,
      env=environ.copy() | { 'OS_PASSWORD' : self.password },
      stdout=asyncio.subprocess.PIPE
    )

    assert(await proc.wait() == 0)
  
  async def get_outputs(self, name):
    trace('Getting stack creation outputs')
    cmd = [
      'openstack',
      '--os-cloud', self.__cloud_name,
      'stack', 'output', 'show',
      '-f', 'json',
      '--all',
      name
    ]
    proc = await asyncio.create_subprocess_exec(
      *cmd,
      stdout=asyncio.subprocess.PIPE,
      env=environ.copy() | { 'OS_PASSWORD' : self.password }
    )

    stdout, _ = await proc.communicate()
    stdout = stdout.decode()
    stdout = json.loads(stdout)
    for k in stdout:
      stdout[k] = json.loads(stdout[k])
    
    return stdout
  
  async def write_hosts_file(self, stack_name):
    outputs = await self.get_outputs(stack_name)
    trace('Writing hosts file')
    with open('ansible/generated_hosts.ini', 'w') as hf:
      hf.write('[manager]\n')
      hf.write(f'{outputs["manager_ip"]["output_value"]} ansible_user=debian ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_python_interpreter=python3\n')
      hf.write('\n')
      hf.write('[worker]\n')
      hf.write(f'{outputs["worker_ip"]["output_value"]} ansible_user=debian ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_python_interpreter=python3\n')
      hf.write('\n')

  def run_ansible(self):
    trace('Running ansible')
    proc = subprocess.Popen([
      'ansible-playbook',
      # Ansible doesn't handle trying to do auth for multiple servers concurrently very well
      # See https://github.com/ansible/ansible/issues/25068
      '--forks', '1',  
      '-i', 'ansible/generated_hosts.ini',
      'ansible/bootstrap.yml'
    ])
    proc.wait()

async def main():
  client = OpenStackClient('nectar-private')
  await client.create_stack('heat/stack.yaml', 'test', {'key_name': 'alaroldai_haliax'})
  if input('Continue? ') != 'y':
    return
  await client.write_hosts_file('test')
  await client.run_ansible()

if __name__ == '__main__':
  asyncio.run(main())
